src/file/writer.rs

```diff
@@ -15,16 +15,290 @@
 // specific language governing permissions and limitations
 // under the License.
 
-use std::io::Write;
+use std::fs::File;
+use std::io::{Seek, SeekFrom, Write};
+use std::rc::Rc;
 
 use basic::PageType;
+use byteorder::{LittleEndian, ByteOrder};
 use column::page::{CompressedPage, Page, PageWriteSpec, PageWriter};
-use errors::Result;
-use file::metadata::ColumnChunkMetaData;
+use column::writer::{ColumnWriter, get_column_writer};
+use errors::{ParquetError, Result};
+use file::{FOOTER_SIZE, PARQUET_MAGIC};
+use file::metadata::*;
+use file::properties::WriterPropertiesPtr;
 use file::statistics::{to_thrift as statistics_to_thrift};
 use parquet_format as parquet;
+use schema::types::{self, SchemaDescriptor, SchemaDescPtr, TypePtr};
 use thrift::protocol::{TCompactOutputProtocol, TOutputProtocol};
-use util::io::Position;
+use util::io::{FileSink, Position};
+
+/// File writer interface.
+pub trait FileWriter {
+  /// Creates new row group from this file writer.
+  fn create_row_group(&mut self) -> Result<&mut RowGroupWriter>;
```

Comment:
The good thing with this is we cannot call `close` until all the returned `RowGroupWriter` are out of scope. The downside is the restrictiveness: now it's hard to pass the `RowGroupWriter` to some struct which lives beyond the current call stack. Not sure if this will be an issue.

Reply:
Yes, I had the same thought, but it looks like this is one of the few options to keep track of row group writers ourselves. Do you have other ideas? Should we just keep a reference count of row group writers and check later when asking for the next one?

Reply:
Yes, I think another approach is to return a unique reference to the `RowGroupWriter`:
```
fn create_row_group(&mut self) -> Result<RowGroupWriter>;
```

And a separate function to explicitly close it:
```
fn close_row_group(&mut self, RowGroupWriter) -> Result<()>;
```

This will call `finalise_row_group_writer` to close the row group writer and set some flag in the file writer indicating that the current row group writer has been closed.  We can do similar things for `RowGroupWriter` to make sure all column writers are handed back before the `close` is called. 

Let me know what you think. This is just some rough thought.

Reply:
Sounds good, I will make changes, see what it looks like, and let you know. Thanks!